pkg/streamingpromql/operators/function_over_range_vector.go

```diff
@@ -22,38 +22,51 @@ import (
 type FunctionOverRangeVector struct {
 	Inner types.RangeVectorOperator
 	Pool  *pooling.LimitingPool
-
-	MetadataFunc functions.SeriesMetadataFunction
-	StepFunc     functions.RangeVectorStepFunction
+	Func  functions.FunctionOverRangeVector
 
 	Annotations *annotations.Annotations
 
+	metricNames        *MetricNames
+	currentSeriesIndex int
+
 	numSteps        int
 	rangeSeconds    float64
 	floatBuffer     *types.FPointRingBuffer
 	histogramBuffer *types.HPointRingBuffer
 
-	expressionPosition posrange.PositionRange
+	expressionPosition   posrange.PositionRange
+	emitAnnotationFunc   functions.EmitAnnotationFunc
+	seriesValidationFunc functions.RangeVectorSeriesValidationFunction
 }
 
 var _ types.InstantVectorOperator = &FunctionOverRangeVector{}
 
 func NewFunctionOverRangeVector(
 	inner types.RangeVectorOperator,
 	pool *pooling.LimitingPool,
-	metadataFunc functions.SeriesMetadataFunction,
-	stepFunc functions.RangeVectorStepFunction,
+	f functions.FunctionOverRangeVector,
 	annotations *annotations.Annotations,
 	expressionPosition posrange.PositionRange,
 ) *FunctionOverRangeVector {
-	return &FunctionOverRangeVector{
+	o := &FunctionOverRangeVector{
 		Inner:              inner,
 		Pool:               pool,
-		MetadataFunc:       metadataFunc,
-		StepFunc:           stepFunc,
+		Func:               f,
 		Annotations:        annotations,
 		expressionPosition: expressionPosition,
 	}
+
+	if f.SeriesValidationFuncFactory != nil {
+		o.seriesValidationFunc = f.SeriesValidationFuncFactory()
+	}
+
+	if f.NeedsSeriesNamesForAnnotations || o.seriesValidationFunc != nil {
```

Comment:
do we need to check `seriesValidationFunc` here? Wouldn't `FunctionOverRangeVector` know to set `NeedsSeriesNamesForAnnotations` if `seriesValidationFunc` needs it?

Reply:
Yep, fair, I'll remove this.